### PR TITLE
infoschema: add plan field to the statement summary tables

### PR DESCRIPTION
**Pull request by djshow832**
### What problem does this PR solve? 
Add plan and plan digest to the statement summary tables.
Same SQLs with different plans are summarized in different records in the tables.
Record the SQL and plan that appear at the first time in each summary, not the last time, to increase performance.

Since same SQL may be split into several records, `max-stmt-count` should increase.

### What is changed and how it works?
Add `plan_digest` and `plan` fields.
`schema_name` + `digest` + `prev_sql_digest` + `plan_digest` are combined as a key of the summary map.
Default value of `max-stmt-count` is increased to 200.
Record the SQL and plan that appear at the first time in each summary, not the last time.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
```
mysql> select * from performance_schema.events_statements_summary_by_digest where plan != ''\G
*************************** 1. row ***************************
       SUMMARY_BEGIN_TIME: 2019-12-22 16:00:00
         SUMMARY_END_TIME: 2019-12-22 16:30:00
                STMT_TYPE: select
              SCHEMA_NAME:
                   DIGEST: ff1ba2b3cf4f7452291642f6399447979825a310ef25a78f01288580f398a3c9
              DIGEST_TEXT: select table_id , is_index , hist_id , distinct_count , version , null_count , tot_col_size , stats_ver , flag , correlation , last_analyze_pos from mysql . stats_histograms where table_id = ?
              TABLE_NAMES: mysql.stats_histograms
              INDEX_NAMES: stats_histograms:tbl
              SAMPLE_USER: NULL
               EXEC_COUNT: 17
              SUM_LATENCY: 17842186
              MAX_LATENCY: 1417064
              MIN_LATENCY: 697887
              AVG_LATENCY: 1049540
        AVG_PARSE_LATENCY: 91016
        MAX_PARSE_LATENCY: 150406
      AVG_COMPILE_LATENCY: 309173
      MAX_COMPILE_LATENCY: 389786
             COP_TASK_NUM: 34
     AVG_COP_PROCESS_TIME: 0
     MAX_COP_PROCESS_TIME: 0
  MAX_COP_PROCESS_ADDRESS: NULL
        AVG_COP_WAIT_TIME: 0
        MAX_COP_WAIT_TIME: 0
     MAX_COP_WAIT_ADDRESS: NULL
         AVG_PROCESS_TIME: 0
         MAX_PROCESS_TIME: 0
            AVG_WAIT_TIME: 0
            MAX_WAIT_TIME: 0
         AVG_BACKOFF_TIME: 0
         MAX_BACKOFF_TIME: 0
           AVG_TOTAL_KEYS: 0
           MAX_TOTAL_KEYS: 0
       AVG_PROCESSED_KEYS: 0
       MAX_PROCESSED_KEYS: 0
        AVG_PREWRITE_TIME: 0
        MAX_PREWRITE_TIME: 0
          AVG_COMMIT_TIME: 0
          MAX_COMMIT_TIME: 0
   AVG_GET_COMMIT_TS_TIME: 0
   MAX_GET_COMMIT_TS_TIME: 0
  AVG_COMMIT_BACKOFF_TIME: 0
  MAX_COMMIT_BACKOFF_TIME: 0
    AVG_RESOLVE_LOCK_TIME: 0
    MAX_RESOLVE_LOCK_TIME: 0
AVG_LOCAL_LATCH_WAIT_TIME: 0
MAX_LOCAL_LATCH_WAIT_TIME: 0
           AVG_WRITE_KEYS: 0
           MAX_WRITE_KEYS: 0
           AVG_WRITE_SIZE: 0
           MAX_WRITE_SIZE: 0
     AVG_PREWRITE_REGIONS: 0
     MAX_PREWRITE_REGIONS: 0
            AVG_TXN_RETRY: 0
            MAX_TXN_RETRY: 0
        SUM_BACKOFF_TIMES: 0
            BACKOFF_TYPES: NULL
                  AVG_MEM: 16672
                  MAX_MEM: 16672
        AVG_AFFECTED_ROWS: 0
               FIRST_SEEN: 2019-12-22 16:18:25
                LAST_SEEN: 2019-12-22 16:19:13
        QUERY_SAMPLE_TEXT: select table_id, is_index, hist_id, distinct_count, version, null_count, tot_col_size, stats_ver, flag, correlation, last_analyze_pos from mysql.stats_histograms where table_id = 43
         PREV_SAMPLE_TEXT:
              PLAN_DIGEST: db05f9e134185031e52539c303d1930de3cf6e5d4881aee7ef97ba87a9a0a6d9
                     PLAN: 	Projection_4    	root	10	mysql.stats_histograms.table_id, mysql.stats_histograms.is_index, mysql.stats_histograms.hist_id, mysql.stats_histograms.distinct_count, mysql.stats_histograms.version, mysql.stats_histograms.null_count, mysql.stats_histograms.tot_col_size, mysql.stats_histograms.stats_ver, mysql.stats_histograms.flag, mysql.stats_histograms.correlation, mysql.stats_histograms.last_analyze_pos
	└─IndexLookUp_10	root	10
	  ├─IndexScan_8 	cop 	10	table:stats_histograms, index:table_id, is_index, hist_id, range:[43,43], keep order:false, stats:pseudo
	  └─TableScan_9 	cop 	10	table:stats_histograms, keep order:false, stats:pseudo
```

Code changes

 - N/A

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release note

 - And `plan` and `plan_digest` to the statement summary tables.
 - Change default value of configuration `max-stmt-count` in [stmt-summary] to 200.

### CODE REVIEW

File: util/stmtsummary/statement_summary.go [hunk 1]
```diff
@@ -52,10 +57,11 @@ type stmtSummaryByDigestKey struct {
 // `prevSQL` is included in the key To distinguish different transactions.
 func (key *stmtSummaryByDigestKey) Hash() []byte {
 	if len(key.hash) == 0 {
-		key.hash = make([]byte, 0, len(key.schemaName)+len(key.digest)+len(key.prevDigest))
+		key.hash = make([]byte, 0, len(key.schemaName)+len(key.digest)+len(key.prevDigest)+len(key.planDigest))
 		key.hash = append(key.hash, hack.Slice(key.digest)...)
 		key.hash = append(key.hash, hack.Slice(key.schemaName)...)
 		key.hash = append(key.hash, hack.Slice(key.prevDigest)...)
+		key.hash = append(key.hash, hack.Slice(key.planDigest)...)
```
Comment (lonng) [inline util/stmtsummary/statement_summary.go]: The same plan but different row count may cause the hash different, is it expected?

Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: No, row count is not in normalized plan, so plan digest are the same.
Refer to https://github.com/pingcap/tidb/blob/master/util/plancodec/codec.go#L274

Comment (lonng) [reply util/stmtsummary/statement_summary.go]: Add some tests to cover this case (different plan but the same plan digest).

Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: Addressed. Refer to https://github.com/pingcap/tidb/pull/14182/commits/00cbf9ec78a19c55b16c85736e9f6b362436d601

File: executor/adapter.go [hunk 2]
```diff
@@ -876,6 +876,9 @@ func (a *ExecStmt) SummaryStmt() {
 	}
 	sessVars.SetPrevStmtDigest(digest)
 
+	plan := plannercore.EncodePlan(a.Plan)
```
Comment (crazycs520) [inline executor/adapter.go]: No  performance impact?

Comment (crazycs520) [reply executor/adapter.go]: But it look ok from the result of benchmark Report...

**Comment by djshow832** [comment]
> LGTM.
> BTW, The plan no need to store the real plan, store the normalized plan is ok. because store the plan need encode plan and will user snappy compress the plan, it will have some performance impact?

I thought about it, but normalized plan is hard to read. E.g.
```
0	37_1	0	table:tidb, index:VARIABLE_NAME
```
I don't know what `0` and `37_1` means.

I'm thinking about printing the plan only at the first time,  but I also want to make fields `query_sample_text` and `plan` are consistent.
So I make a benchmark, which shows it affects only a little performance degression.

File: executor/adapter.go (hunk 2 above)
Comment (djshow832) [reply executor/adapter.go]: I changed it to recording the plan only at the first time.

File: util/stmtsummary/statement_summary.go [hunk 3]
```diff
@@ -601,8 +601,26 @@ func (ssbd *stmtSummaryByDigest) collectHistorySummaries(historySize int) []*stm
 }
 
 func newStmtSummaryByDigestElement(sei *StmtExecInfo, beginTime int64, intervalSeconds int64) *stmtSummaryByDigestElement {
+	// sampleSQL / sampleUser / samplePlan / prevSQL / indexNames store the values shown at the first time,
+	// because it compacts performance to update every time.
+	maxSQLLength := config.GetGlobalConfig().StmtSummary.MaxSQLLength
+	sampleSQL := sei.OriginalSQL
+	if len(sampleSQL) > int(maxSQLLength) {
+		// Make sure the memory of original `sampleSQL` will be released.
+		sampleSQL = string([]byte(sampleSQL[:maxSQLLength]))
```
Comment (lonng) [inline util/stmtsummary/statement_summary.go]: Do we need to attach the information of sample query total length, e.g: `select * from tt (total: 2304bytes)`

Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: Addressed.

**Pull request merged**